### ISSUE TITLE
Community expansion: steps to being welcoming and inclusive

### ISSUE THREAD

**Issue by blahah** [BioJulia/Bio.jl#102]
We've talked on gitter a bunch of times about this stuff. I'm putting it here so it's more visible, and actionable.

The basic problem is this: open source communities tend to have very low diversity, and to be offputting to a large proportion of potential contributors. As in all things, we should try to do better.

Some background from the open science/open source communities:
- http://redmonk.com/dberkholz/2012/10/29/githubs-success-and-low-barriers-to-entry/
- http://mozillascience.github.io/leadership-training
- http://producingoss.com/en/producingoss.html (this is especially good and thorough)

These are some concrete steps we can take to make the project more welcoming and inclusive:
- [x] Have a credible, enforcible code of conduct (we currently follow the [Julia community standards](http://julialang.org/community/standards/), but these are not particularly strong)
  - [ ] with specific routes for reporting bad behaviour, including a way to hold group leaders or core developers accountable (discussed before - options include asking OBF board, Mozilla Science)
  - [x] make it clear we **always** act on reports, and do so openly
- [x] Make a [contributing guide](https://github.com/BioJulia/Bio.jl/blob/master/CONTRIBUTING.md)
- [ ] Make contributing trivially easy
  - [x] Link to the contributing guide prominently on the website
  - [ ] Have a very clear guide to setting up a development environment.
  - [x] Be clear that we are here to support people who want to contribute. Have a list of people willing to mentor contributors.
- [ ] Be super, uber way over-the-top friendly in our welcome text. Make it clear that we want people, **all** kinds of people, to be involved and value their contribution.
  - [ ] Make it clear that lurking is encouraged! Watch us collaborate, we hope you might eventually want to get involved, but just watching is totally fine.
  - [ ] Make it clear that we care a lot about quality, but we make (lots of) mistakes. If someone sees a typo, or mistake however trivial, we want to encourage them to tell us or better, submit a PR.
  - [x] Every contributor should get credit. In a contributors file, and in release notes. And in publications. But people can of course request to remain anonymous.
- [ ] Have a policy to avoid private discussions. If discussions have to be had in private (e.g. because they are by email) they should be copied in public as well.
- [x] Practise conspicuous code review
  - [ ] have a code review policy
  - [ ] make it highly visible and transparent
- [ ] Avoid any one person having control over the project
  - [ ] Ensure that organisation assets are held in common, but securely if necessary
  - [ ] Place some backup control in the hands of an external power (e.g. the OBF)
- [ ] Avoid complexity where possible. If it's necessary, go to whatever lengths it takes to reduce it and make it accessible.

What have I missed?

**Comment by Blahah** [comment]
Also, I hope this will be a kind of 'living issue' - that we will never close it. It's vanishingly unlikely that we will have done all we can. Having this issue is a signal that we want to do more. So let's update it as we make progress, but keep open to new ideas.

**Comment by Ward9250** [comment]
Perhaps a contributiors.JSON, when you make a PR, you get to add yourself and some basic profile info to it. People added could then be displayed dynamically on the website.

**Comment by Ward9250** [comment]
I think for a lot of these things - like the code review, and other guidelines and policies, can be added to the Judo generated website, now hosted on the gh-pages. That way they are in the source, and online. Unless we want a seperate website for Bio.jl itself, and then BioJulia the organisation.

**Comment by dcjones** [comment]
We'd just have to move `CONTRIBUTING.md` to the `doc` directory and it will automatically be included.

**Comment by Ward9250** [comment]
Very cool. I've been readin the GitBook @Blahah links at the top, by Mozilla, they suggest that codes of conduct, and contributing.md are well placed in git repo's of their own, and then added to all other repositories where relevant, by using git submodules - I think Angular do this.

**Comment by jelena121** [comment]
I think a friendly website with a clear procedure for joining and contributing, and clear community guidelines, goes a long way. I've always been curious about open source programming projects, but have never joined any, because I wasn't sure how to find or join them. Actually, it's still a bit of a mystery to me. 

Also making sure that new contributors are actually welcomed, and that discussions are friendly/cooperative rather than antagonistic. Seeing snarky comments, patronising comments, "Let me google that for you", etc, makes me leave forums straight away even if they're directed at people other than me, because it signals that it's not the sort of community that I particularly want to be a part of, so I might as well not waste time on it. I guess that ties into the code of conduct / community guidelines. 

This is from an event rather than an open source project, but 9 Worlds has a great track record of both putting in place and enforcing anti-harrassment policies, linking in case it's useful: https://nineworlds.co.uk/anti-harassment-policy
I think something similar could be repurposed, with less focus on physical space, but things like oppressive language are still very relevant.  

Finally, actively encouraging a diverse range of people to join, and making sure they feel welcome when they do.

**Comment by Blahah** [comment]
@jelena121 thank you for this, and welcome to BioJulia :)

It's great to have your perspective as someone who has a lot to offer open source projects, but hasn't been involved yet. Your experience highlights the problem perfectly.

Based on what you said, I've added the following to the checklist:

- [ ] Prominently feature a clear procedure for joining and contributing on the website and link to it from all project entry points
- [ ] Have a policy about language used in discussion and comments - friendly/cooperative and respectful language is expected, while rude, belittling or aggressive language will be removed. This in addition to any rules about discrimination etc.

The NineWorlds policy looks really good, and if they have a good reputation *and* empirically it's effective then we should learn from them. When I have some time to focus on it (screaming baby on lap now), I'll see if I can distil what they have that we're missing.

It would also be great if we could find out from more people why they *didn't* get involved. I wonder if we could have something like, at the end of the contributing guide: "If you wanted to contribute, but we've put you off in any way, we'd love to learn what we can do better. You can send us an anonymous message: " and then a simple form. Thoughts on whether this is a good idea?

In addition, we should start discussion how to help a diverse range of people find the project in the first place.

**Comment by Blahah** [comment]
@Ward9250 I Iike the idea of inviting contributors to add themselves to the `.json` with their first PR. If a new contributor makes a PR and *doesn't* add themselves, we could suggest it in code review (in case they didn't realise/forgot). Having it dynamic on the website is easy, we can just ajax load it from github.

**Comment by prcastro** [comment]
As Jelena pointed, I have seen a lot of things on OSS communities that would put someone off. Specially if you're not from the same social group that the community is composed of (white caucasian middle-to-upper-class men, usually). All these suggestions could prevent this from happening in BioJulia.

I believe this kind of policy really help creating a more diverse, inclusive, fun and healthy community to contribute with. I'm really looking forward to help with anything needed to make this happen.

**Comment by Blahah** [comment]
I just read this blog post: http://sarah.thesharps.us/2015/10/06/what-makes-a-good-community/

It is excellent - it talks about different levels of community that must be built to support different aspects of inclusiveness and diversity. Again when I have time, I'll try to distil some things from it into our action plan.

**Comment by dcjones** [comment]
That is a good post, and by someone who has a lot of experience with an occasionally toxic community.

By the way, JuliaGeo has a [nice page for new contributors](https://github.com/JuliaGeo/meta/wiki/How-to-get-Involved). One easy step for us would be to mimic that and just have a page explaining in practical terms how to follow development and get involved, beyond just rules and policies which are laid out already in CONTRIBUTING.

**Comment by Ward9250** [comment]
I very much like how JuliaGeo have described the registration of the package under the org.

**Comment by Blahah** [comment]
That JuliaGeo page is good - and as you say we can easily dupe that. +1

I'm also leaving another link here. Haven't had time to read it yet, but it's about OpenNews' experience with Codes of Conduct: http://incisive.nu/2014/codes-of-conduct/ (via @kaythaney)

**Comment by Blahah** [comment]
I saw the [rOpenSci Onboarding](https://github.com/ropensci/onboarding) page today for the first time, and it's very nice. They have a clear procedure for bringing new packages and developers in, and it is explained well. They have nice policies and by following back to the source I found this [very well though-out code of conduct](http://citizencodeofconduct.org/) from which they take the definition of harassment.
